Refactor and implement `Restart-Computer` for `Un*x` and macOS

brendandburns opened the pull request:
# PR Summary
Implement the `Restart-Computer` command for Un*x and MacOS

## PR Context
Compliment to the recent implementation of `Stop-Computer` (#11151)

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
- **User-facing changes**
    - [X] Not Applicable
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.

brendandburns committed:
Refactor and implement Restart-Computer

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs:
```diff
@@ -10,22 +10,70 @@
+    /// Cmdlet to restart computer.
+    /// </summary>
+    [Cmdlet(VerbsLifecycle.Restart, "Computer", SupportsShouldProcess = true,
+        HelpUri = "https://go.microsoft.com/fwlink/?LinkID=135263", RemotingCapability = RemotingCapability.SupportedByCommand)]
```
Please use new value for `LinkID=135263`

brendandburns replied:
Done.

iSazonov commented:
@brendandburns Please don't rebase until maintainers ask you. Rebasing complicates reviewing.

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs:
```diff
@@ -10,22 +10,70 @@
+        /// </summary>
+        protected override void BeginProcessing()
+        {
+            var args = "";
```
Perhaps initialize `args = "-P now"` to optimize for Linux and just check for macOS?

brendandburns replied:
done.

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs:
```diff
@@ -10,22 +10,70 @@
+            {
+                args = "now";
+            }
+            RunCommand("/sbin/shutdown", args);
```
The style rules we have expect a newline after closing brace
```suggestion

            RunCommand("/sbin/shutdown", args);
```

iSazonov replied:
@brendandburns Please address the comment.

brendandburns replied:
Done.

SteveL-MSFT reviewed (changes requested):
The Windows Computer cmdlets already use a test hook for testing, should use the same thing here.  See https://github.com/PowerShell/PowerShell/blob/master/src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs#L2130

iSazonov commented:
We merged #11151 without new tests.

brendandburns commented:
I will address comments and add tests for both commands.

Thanks

brendandburns commented:
@iSazonov is there a preferred style of addressing comments for the PowerShell repo? Should I just add additional commits instead of squashing commits?

(I'm new here :)

Thanks

iSazonov commented:
> Should I just add additional commits instead of squashing commits?

Yes, please add new commits. Some commits or one for all comments.

SteveL-MSFT commented:
@brendandburns unless there is a need to preserve history, we always squash on merge.  For the sake of the reviewers, it's always been easier to add new commits so we can see what exactly changed.

brendandburns committed:
Add a test for Stop-Computer

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Stop-Computer.Tests.ps1:
```diff
@@ -10,7 +10,7 @@ $DefaultResultValue = 0
 {
     # set up for testing
-    $PSDefaultParameterValues["it:skip"] = ! $IsWindows
+    # $PSDefaultParameterValues["it:skip"] = ! $IsWindows
```
Please remove the commented code.

brendandburns replied:
Done.

brendandburns committed:
Address changes.

brendandburns commented:
@iSazonov @SteveL-MSFT 

Sorry it took so long for me to get back to this. I've added tests for both `Restart-Computer` and `Stop-Computer` on Unix.

Please take another look.

Thanks!

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs:
```diff
@@ -6,26 +6,97 @@
+    }
+
+    /// <summary>
+    /// A base class for cmdlets that can run shell commands
```
```suggestion
    /// A base class for cmdlets that can run shell commands.
```

iSazonov commented on src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs:
```diff
@@ -67,21 +129,15 @@ protected override void StopProcessing()
-
-        private void doShutdown() {
+        /// <summary>
+        /// Run a command
```
```suggestion
        /// Run a command.
```

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1:
```diff
@@ -30,12 +29,18 @@ try
         It "Should support -computer parameter" {
+            if ( ! $IsWindows ) {
+                return
+            }
```
Please use `It "Should support -computer parameter" -Skip:(!$IsWindows) {`

brendandburns replied:
done for all.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1:
```diff
@@ -30,12 +29,18 @@ try
         It "Should support WsmanAuthentication types" {
+            if ( ! $IsWindows ) {
+                return
+            }
```
The same.

brendandburns replied:
done for all.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1:
```diff
@@ -46,6 +51,9 @@ try
         It "Should wait for a remote system" {
+            if ( ! $IsWindows ) {
+                return
+            }
```
The same.

brendandburns committed:
Update src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs

Co-Authored-By: Ilya <darpa@yandex.ru>

brendandburns committed:
Update src/Microsoft.PowerShell.Commands.Management/commands/management/ComputerUnix.cs

Co-Authored-By: Ilya <darpa@yandex.ru>

brendandburns committed:
Address comments.

brendandburns commented:
@iSazonov comments addressed, please re-check.

Thanks!

iSazonov commented:
@brendandburns Did you address all @SteveL-MSFT 's comments?

brendandburns commented:
yes, I verified in the code that I had addressed both of the comments from @SteveL-MSFT 

Thanks

iSazonov commented:
@SteveL-MSFT Please update your review.

PoshChan commented:
@TravisEz13, this is the reminder you requested 1 hour ago

iSazonov commented:
@brendandburns Thanks for your contribution!
